pkg/admission/customresourcevalidation/network/validate_network.go

```diff
@@ -0,0 +1,130 @@
+package network
+
+import (
+	"fmt"
+	"io"
+	"net"
+
+	"k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/admission"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/origin/pkg/admission/customresourcevalidation"
+)
+
+const PluginName = "config.openshift.io/ValidateNetwork"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return customresourcevalidation.NewValidator(
+			map[schema.GroupResource]bool{
+				configv1.Resource("networks"): true,
+			},
+			map[schema.GroupVersionKind]customresourcevalidation.ObjectValidator{
+				configv1.GroupVersion.WithKind("Network"): networkV1{},
+			})
+	})
+}
+
+func toNetworkV1(uncastObj runtime.Object) (*configv1.Network, field.ErrorList) {
+	if uncastObj == nil {
+		return nil, nil
+	}
+
+	allErrs := field.ErrorList{}
+
+	obj, ok := uncastObj.(*configv1.Network)
+	if !ok {
+		return nil, append(allErrs,
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"Network"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"config.openshift.io/v1"}))
+	}
+
+	return obj, nil
+}
+
+type networkV1 struct {
+}
+
+// This only validates the syntax; the operator will worry about the semantics
+func validateNetworkSpec(spec configv1.NetworkSpec) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(spec.ClusterNetwork) == 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork"), nil, "must have at least 1 entry"))
+	}
+	for i, cnet := range spec.ClusterNetwork {
+		_, cidr, err := net.ParseCIDR(cnet.CIDR)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork").Index(i).Child("cidr"), cnet.CIDR, err.Error()))
+		}
+		_, bits := cidr.Mask.Size()
+		if cnet.HostPrefix > uint32(bits) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork").Index(i).Child("hostPrefix"), cnet.HostPrefix, "too large for address type"))
+		}
+	}
+
+	if len(spec.ServiceNetwork) != 1 {
```

Comment:
If there can only be one, why is this a list?

List aside, checking for 0 and >1 separately will allow for better error messages.

Reply:
Eventually we will support multiple CIDRs (eg, one IPv4 and one IPv6 network, but there are other use cases too). But k8s 1.12 only supports a single CIDR so we can reject anything else for now.